### ISSUE TITLE
Computed Variables for Algorithm Outputs are NULL in Graphical Modeler

### ISSUE THREAD

**Issue by StefanBrand** [qgis/QGIS#45865]
### What is the bug or the crash?

I have tried to use the computed variables for an algorithm output (e.g. `@Reproject_layer_OUTPUT_minx `) in a `Pre-calculated value` field in a subsequent algorithm. All but those coming from the input layer are always NULL.

### Steps to reproduce the issue

1. Start Graphical Modeler
2. Add Input "Vector Layer"
3. Add "Reproject layer" (or other tools)
4. Add "Create grid"
5. For "Grid extent" choose "Pre-calculated value" and paste
```
to_string( @Reproject_layer_OUTPUT_minx ) || ',' ||
to_string( @Reproject_layer_OUTPUT_maxx ) || ',' ||
to_string( @Reproject_layer_OUTPUT_miny ) || ',' ||
to_string( @Reproject_layer_OUTPUT_maxy )
```

Output: 

```
Input Parameters:
{ CRS: QgsCoordinateReferenceSystem('EPSG:32634'), EXTENT: None, HOVERLAY: 0, HSPACING: 24000, OUTPUT: 'TEMPORARY_OUTPUT', TYPE: 2, VOVERLAY: 0, VSPACING: 24000 }
Horizontal spacing is too large for the covered area.
Error encountered while running Create grid
Execution failed after 0.12 seconds
```

6. Trying again with the following works:
```
to_string(@CountryLayer_minx ) || ', ' ||
to_string( @CountryLayer_maxx ) || ', ' ||
to_string( @CountryLayer_miny ) || ', ' ||
to_string(  @CountryLayer_maxy )
```

Output:

```
{ CRS: QgsCoordinateReferenceSystem('EPSG:32634')', EXTENT: '4744834.27221361, 5402446.1378068905, 4125728.6944347215, 5306091.750087504', HOVERLAY: 0, HSPACING: 24000, OUTPUT: 'memory:Grid', TYPE: 2, VOVERLAY: 0, VSPACING: 24000 }
```

### Versions

<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0//EN" "http://www.w3.org/TR/REC-html40/strict.dtd">
<html><head><meta http-equiv="Content-Type" content="text/html; charset=utf-8" /><style type="text/css">
p, li { white-space: pre-wrap; }
</style></head><body>

QGIS version | 3.22.0-Białowieża | QGIS code revision | d9022691f1
-- | -- | -- | --
Qt version | 5.12.8
Python version | 3.8.10
GDAL/OGR version | 3.0.4
PROJ version | 6.3.1
EPSG Registry database version | v9.8.6 (2020-01-22)
Compiled against GEOS | 3.8.0-CAPI-1.13.1 | Running against GEOS | 3.8.0-CAPI-1.13.1
SQLite version | 3.31.1
PDAL version | 2.0.1
PostgreSQL client version | 12.8 (Ubuntu 12.8-0ubuntu0.20.04.1)
SpatiaLite version | 4.3.0a
QWT version | 6.1.4
QScintilla2 version | 2.11.2
OS version | Ubuntu 20.04.3 LTS
  |   |   |  
Active Python plugins
valuetool | 3.0.8
quick_map_services | 0.19.11.1
zoom_level | 0.1
QuickWKT | 3.1
processing | 2.12.99
db_manager | 0.1.20
grassprovider | 2.12.99
sagaprovider | 2.12.99
MetaSearch | 0.3.5

</body></html>

### Supported QGIS version

- [X] I'm running a supported QGIS version according to the roadmap.

### New profile

- [ ] I tried with a new QGIS profile

### Additional context

Test QGZ project: [test.qgz.zip](https://github.com/qgis/QGIS/files/7473540/test.qgz.zip)

The generated Python script:

```python
"""
Model exported as python.
Name : test_out_grid
Group : 
With QGIS : 31803
"""

from qgis.core import QgsProcessing
from qgis.core import QgsProcessingAlgorithm
from qgis.core import QgsProcessingMultiStepFeedback
from qgis.core import QgsProcessingParameterVectorLayer
from qgis.core import QgsProcessingParameterFeatureSink
from qgis.core import QgsProcessingParameterBoolean
from qgis.core import QgsCoordinateReferenceSystem
from qgis.core import QgsExpression
import processing


class Test_out_grid(QgsProcessingAlgorithm):

    def initAlgorithm(self, config=None):
        self.addParameter(QgsProcessingParameterVectorLayer('CountryLayer', 'Country Layer', types=[QgsProcessing.TypeVectorPolygon], defaultValue=None))
        self.addParameter(QgsProcessingParameterFeatureSink('Out_grid', 'out_grid', type=QgsProcessing.TypeVectorPolygon, createByDefault=True, defaultValue=None))
        self.addParameter(QgsProcessingParameterBoolean('VERBOSE_LOG', 'Verbose logging', optional=True, defaultValue=False))

    def processAlgorithm(self, parameters, context, model_feedback):
        # Use a multi-step feedback, so that individual child algorithm progress reports are adjusted for the
        # overall progress through the model
        feedback = QgsProcessingMultiStepFeedback(2, model_feedback)
        results = {}
        outputs = {}

        # Create grid
        alg_params = {
            'CRS': QgsCoordinateReferenceSystem('EPSG:32634'),
            'EXTENT': QgsExpression('to_string(@Reproject_layer_OUTPUT_minx ) || \', \' ||\nto_string( @Reproject_layer_OUTPUT_maxx ) || \', \' ||\nto_string( @Reproject_layer_OUTPUT_miny ) || \', \' ||\nto_string(  @Reproject_layer_OUTPUT_maxy )').evaluate(),
            'HOVERLAY': 0,
            'HSPACING': 24000,
            'TYPE': 2,
            'VOVERLAY': 0,
            'VSPACING': 24000,
            'OUTPUT': parameters['Out_grid']
        }
        outputs['CreateGrid'] = processing.run('native:creategrid', alg_params, context=context, feedback=feedback, is_child_algorithm=True)
        results['Out_grid'] = outputs['CreateGrid']['OUTPUT']

        feedback.setCurrentStep(1)
        if feedback.isCanceled():
            return {}

        # Reproject layer
        alg_params = {
            'INPUT': parameters['CountryLayer'],
            'OPERATION': '',
            'TARGET_CRS': QgsCoordinateReferenceSystem('EPSG:32634'),
            'OUTPUT': QgsProcessing.TEMPORARY_OUTPUT
        }
        outputs['ReprojectLayer'] = processing.run('native:reprojectlayer', alg_params, context=context, feedback=feedback, is_child_algorithm=True)
        return results

    def name(self):
        return 'test_out_grid'

    def displayName(self):
        return 'test_out_grid'

    def group(self):
        return ''

    def groupId(self):
        return ''

    def createInstance(self):
        return Test_out_grid()

```

**Comment by StefanBrand** [comment]
Cannot reproduce

**Comment by StefanBrand** [comment]
Reopening because it stopped working after restarting QGIS

**Comment by StefanBrand** [comment]
Observation: I added the "Reproject layer" algorithm to the dependencies of the "Create grid" algorithm and it started working again. This feels like a race condition.

Is the dependency not implied if I use a computed variable in a parameter field of a later algorithm?

**Comment by aldenln** [comment]
I'm also having a problem with the variables calculated in the model builder.

I tried to use the variables calculated by the algorithms ('feature filter' and 'export layer information' ). I used them in the 'conditional arms' and 'increase notice'. I have tried to extract any information from them but the answer is either null or results in an interruption of the algorithm.

My processing filters out 'ok' and 'problem' features arranged in hundreds of layers. Not all layers have 'problem' features so I want to save the problem features in new files only if they exist. For this I use the 'conditional arms' and want to evaluate if the OUTPUT of the 'feature filter' contains features. Without success, the return is always 'Failed condition'.

I started testing in different ways using the 'increase warning' to get information in the log. I couldn't get any information from these calculated variables. When I did the same tests with the model's input variables, they worked well.

Going beyond the delimited theme here, I apologize if it's not pertinent, on another occasion I tried using the 'set project variable' algorithm to store custom names for output files. Something trivial, 'base_name'+timestamp stored in a variable, which was then passed as the filename to the output.

In the first run of this model, a failure occurs, the value passed is null. On subsequent runs, the model works, but the value that is passed to the output is the name calculated in the previous run. As if the value were updated after execution.

I was wondering if this error could not be related to this issue. Let the experts speak!

I tested it with Qgis 3.16.4-1 then used a fresh install, currently 3.20.3:

Código da versão do QGIS: 495fbaec
Qt version: 5.15.2
Python version: 3.9.5
GDAL/OGR version: 3.3.1
PROJ version: 8.1.1
EPSG Registry database version: v10.028 (2021-07-07)
GEOS version: 3.9.1-CAPI-1.14.2
SQLite version: 3.35.2
PDAL version: 2.3.0
PostgreSQL client version: 13.0
SpatiaLite version: 5.0.1
QWT version: 6.1.3
QScintilla2 version: 2.11.5
OS version: Windows 10 Version 2009

Active Python plugins: QuickOSM, quick_map_services, db_manager, processing

<>
